### PR TITLE
Support XPU for auto-paralllel LLaMa

### PR DESCRIPTION
**Pull request by From00**
### PR types
New features

### PR changes
Models

### Description

Llama模型适配xpu自动并行训练，目前仅支持动态图+纯dp（只包含allreduce通信）。
依赖主框架PR：https://github.com/PaddlePaddle/Paddle/pull/70997

### CODE REVIEW

File: paddlenlp/transformers/llama/modeling_auto.py [hunk 1]
```diff
@@ -935,7 +955,22 @@ def _prepare_decoder_attention_mask(attention_mask, input_shape, past_key_values
...
+            expanded_attn_mask = paddle.where(expanded_attn_mask.cast("bool"), x, y).astype(dtype)
+        elif get_env_device() == "xpu":
+            x = paddle.to_tensor(0.0, dtype="float32")
+            y = paddle.to_tensor(-1.7005809656952787e38, dtype="float32")
+            expanded_attn_mask = paddle.where(expanded_attn_mask.cast("bool"), x, y)
+        elif get_env_device() == "gcu":
+            min_val = paddle.finfo(dtype).min
+            x = paddle.to_tensor(0.0, dtype=dtype)
+            y = paddle.to_tensor(min_val, dtype=dtype)
+            expanded_attn_mask = paddle.where(expanded_attn_mask.cast("bool"), x, y).astype(dtype)
```
Comment (zhiqiu) [inline paddlenlp/transformers/llama/modeling_auto.py:953]: what's mask generation differs between different devices.

Comment (From00) [reply paddlenlp/transformers/llama/modeling_auto.py:953]: The mask generation logic is same as here: https://github.com/PaddlePaddle/PaddleNLP/blob/develop/paddlenlp/transformers/llama/modeling.py#L1606. 

For the following two reasons that XPU needs a different mask:
1. The flash_attention kernel implemented in XPU is different than in GPU, which may lead to numeric overflow when the mask value is too small. Therefore, a specific mask number `-1.7005809656952787e38` is needed. @runzhech is fixing this issue and we can use `paddle.finfo(dtype).min` like GPU after fixed.
2. The flash_attention kernel in XPU requires the mask input to be `float32`，so the `astype(dtype)` cannot be added in XPU mask generation.

See these two PRs for more details: https://github.com/PaddlePaddle/PaddleNLP/pull/9495, https://github.com/PaddlePaddle/PaddleNLP/pull/9652

**Review by zhiqiu — DISMISSED** [review]
LGTM

**Review by zhiqiu — APPROVED** [review]
LGTM

File: paddlenlp/transformers/llama/modeling_auto.py [hunk 2]
```diff
@@ -308,7 +311,7 @@ def __init__(self, config: LlamaConfig, layerwise_recompute: bool = False, ipp:
         self.ipp = ipp
 
         self.use_fused_rope = config.use_fused_rope
-        if self.use_fused_rope:
+        if self.use_fused_rope and get_env_device() not in ["npu", "mlu", "xpu", "gcu", "intel_hpu"]:
```
Comment (wawltor) [inline paddlenlp/transformers/llama/modeling_auto.py:314]: 自动并行 modeling_network.py 和 modeling_auto.py的关系是什么，modeling_network.py需要同步修改吗？

Comment (From00) [reply paddlenlp/transformers/llama/modeling_auto.py:314]: modeling_network.py是中层API实现，modeling_auto.py是基础API实现，这两套理想状态下应该是合成一份的。这个PR只是昆仑适配的第一步，目前只支持动半纯dp的场景，后边还需要继续测试和迭代，等这块完善之后，后续应该会有专门的线条将 modeling_network.py和modeling_auto.py做合并。
